src/wp-includes/link-template.php

```diff
@@ -1871,11 +1874,49 @@ function get_adjacent_post( $in_same_term = false, $excluded_terms = '', $previo
 			}
 
 			$where .= ' AND tt.term_id IN (' . implode( ',', $term_array ) . ')';
+			$tax_query[] = array(
+				'taxonomy' => $taxonomy,
+				'terms'    => $term_array,
+				'field'    => 'term_id',
+				'operator' => 'IN',
+			);
 		}
 
 		if ( ! empty( $excluded_terms ) ) {
 			$where .= " AND p.ID NOT IN ( SELECT tr.object_id FROM $wpdb->term_relationships tr LEFT JOIN $wpdb->term_taxonomy tt ON (tr.term_taxonomy_id = tt.term_taxonomy_id) WHERE tt.term_id IN (" . implode( ',', array_map( 'intval', $excluded_terms ) ) . ') )';
+			$tax_query[] = array(
+				'taxonomy' => $taxonomy,
+				'terms'    => $excluded_terms,
+				'field'    => 'term_id',
+				'operator' => 'NOT IN',
+			);
+		}
+	}
+
+	if ( ! has_filter( "get_{$adjacent}_post_join" ) && ! has_filter( "get_{$adjacent}_post_where" ) && ! has_filter( "get_{$adjacent}_post_sort" ) ) {
```

Comment:
@peterwilsoncc I think this would be the only way to use WP_Query here to be completely safe.

Reply:
After doing this, I would we should deprecate the filters below. That I am unclear on.

Reply:
> @peterwilsoncc I think this would be the only way to use WP_Query here to be completely safe.

I was thinking about this overnight and wondering if the initial approach might be the safest.

It seems risky to fork the code like this and modyfying the request without modifying the clauses introduces other complexities.

Reply:
I think adding caching and using WP_Query are too different problems. I think we want to use WP_Query, then we do this in another ticket. I am going to close PR for now.